pkg/controller/service_instance.go

```diff
@@ -0,0 +1,183 @@
+			return nil, errors.New("Failure retrieving ServiceInstance spec")
+		}
+		previousEncodedChecksum = spec.ObjectMeta.Annotations[secretParametersChecksumAnnotation]
+		updateCount = spec.Spec.UpdateRequests
+	}
+
+	checkSum, err := st.calculateNewCheckSum(&instanceSpec, namespace, previousEncodedChecksum, updateCount)
+	if err != nil {
+		return nil, errors.Wrapf(err, "Failed to generate new checksum for spec %q", spec.GetName())
+	}
+
+	if previousEncodedChecksum != "" && checkSum != previousEncodedChecksum {
+		instanceSpec.Spec.UpdateRequests = updateCount + 1
+	}
+
+	instanceSpec = setAnnotation(instanceSpec, checkSum)
+
+	unstructuredSpec, err := util.RuntimeToUnstructured(&instanceSpec)
+	if err != nil {
+		return nil, errors.Wrapf(err, "Failed to marshal back into unstructed %q", spec.GetName())
+	}
+
+	return unstructuredSpec, nil
+}
+
+func (st *resourceSyncTask) calculateNewCheckSum(instanceSpec *sc_v1b1.ServiceInstance, namespace string, previousEncodedChecksum string, updateCount int64) (string, error) {
+	checksumPayload, err := st.generateSecretParametersChecksumPayload(&instanceSpec.Spec, namespace)
+	if err != nil {
+		return "", err
+	}
+
+	if previousEncodedChecksum != "" {
+		previousCheckSum, err := decodeChecksum(previousEncodedChecksum)
+		if err != nil {
+			return "", errors.Wrapf(err, "Failed to decode checksum for %q", instanceSpec.Name)
+		}
+		if validateChecksum(previousCheckSum, checksumPayload) {
+			return previousEncodedChecksum, nil
+		}
+	}
+	checksum, err := generateChecksum(checksumPayload)
+	if err != nil {
+		return "", err
+	}
+
+	return encodeChecksum(checksum), nil
+}
+
+func (st *resourceSyncTask) generateSecretParametersChecksumPayload(spec *sc_v1b1.ServiceInstanceSpec, namespace string) ([]byte, error) {
+	var buf bytes.Buffer
+
+	for _, parametersFrom := range spec.ParametersFrom {
+		secretKeyRef := parametersFrom.SecretKeyRef
+		secretObj, exists, err := st.store.Get(core_v1.SchemeGroupVersion.WithKind("Secret"), namespace, secretKeyRef.Name)
+		if err != nil {
+			return nil, errors.Wrapf(err, "Failure retrieving secret %q referenced from spec.ParametersFrom", secretKeyRef.Name)
+		}
+		if !exists {
+			return nil, errors.Errorf("Secret %q referenced from spec.ParametersFrom not found", secretKeyRef.Name)
+		}
+		secret, ok := secretObj.(*core_v1.Secret)
+		if !ok {
+			return nil, errors.Errorf("Failure casting secret %q referenced from spec.ParametersFrom", secretKeyRef.Name)
+		}
+
+		secretData := secret.Data[secretKeyRef.Key]
+		if secretData == nil {
+			return nil, errors.Errorf("key %q not found in Secret %q", secretKeyRef.Key, secretKeyRef.Name)
+		}
+		// Using SHA256 here is fine since we will hash the final result with bcrypt anyway
+		hash := sha256.Sum256(secretData)
+		_, err = buf.Write(hash[:])
+		if err != nil {
+			return nil, err
+		}
+	}
+	return buf.Bytes(), nil
+}
+
+func setAnnotation(instanceSpec sc_v1b1.ServiceInstance, checksum string) sc_v1b1.ServiceInstance {
```

Comment:
Just take a pointer an mutate in place? No need to copy it. And actually shallow struct copy does not copy the contents of the map anyway so the original map is mutated here (unless it is `nil`).